### ISSUE TITLE
PQ_INPUTS and QT_INPUTS crash for cubic mixtures.

### ISSUE THREAD

**Issue by JonWel** [CoolProp/CoolProp#1127]
It must be caused by a very recent change, this input combination was still working recently with low level interface.
```Julia
julia> using CoolProp

julia>  PropsSI("Dmolar", "P", 1e5, "T", 400, "PR::R125[0.17956]&R134a[0.43933]&R32[0.38111]")
30.247210923254897

julia> PropsSI("Dmolar", "P", 1e5, "Q", 0, "HEOS::R125[0.17956]&R134a[0.43933]&R32[0.38111]")
15977.479239072072

julia> PropsSI("Dmolar", "P", 1e5, "Q", 0, "PR::R125[0.17956]&R134a[0.43933]&R32[0.38111]")

Please submit a bug report with steps to reproduce this fault, and any error mes
sages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x7fedcc32494 -- unknown function (ip:
000007FEDCC32494)
unknown function (ip: 000007FEDCC32494)
unknown function (ip: 000007FEDCC1D1CC)
unknown function (ip: 000007FEDCC1DDBC)
unknown function (ip: 000007FEDCCAD16D)
unknown function (ip: 000007FEDCC3FA26)
unknown function (ip: 000007FEDCC7848D)
unknown function (ip: 000007FEDCC1D516)
unknown function (ip: 000007FEDCC1E5A3)
unknown function (ip: 000007FEDCCEFD41)
unknown function (ip: 000007FEDCCF037F)
unknown function (ip: 000007FEDCCEEA3A)
PropsSI at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\CoolProp.DLL (unknown line)
PropsSI at C:\Users\s161576\.julia\v0.4\CoolProp\src\CoolProp.jl:34 
jlcall_PropsSI_1402 at  (unknown line)
jl_apply_generic at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll (unknown line)
jl_interpret_toplevel_expr at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\lib julia.dll (unknown line)
jl_interpret_toplevel_thunk_with at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll (unknown line)
jl_eval_with_compiler_p at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll (unknown line)
jl_f_tuple at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll (unknown line)
eval_user_input at REPL.jl:62
jlcall_eval_user_input_1278 at  (unknown line)
jl_apply_generic at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll(unknown line)
anonymous at REPL.jl:92
jl_unprotect_stack at C:\Users\s161576\AppData\Local\Julia-0.4.5\bin\libjulia.dll (unknown line)

C:\Users\s161576>
```
Or a simpler example:
```Julia
julia> PropsSI("Dmolar", "P", 1e5, "Q", 0, "PR::ethane[0.5]&propane[0.5]")
```
Pure fluids doesn't seem affected.

The problem is the same for low level interface (that's with low level interface call that I experienced it first, I used the high level interface here to have more concise example).

**Comment by ibell** [comment]
Hmm... I don't think I did anything that would have messed up the mixtures
with cubic EOS.  I'll investigate.

**Comment by JonWel** [comment]
Seems to be an "out of range" problem.
In the VLERoutines file, at [this line](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Helmholtz/VLERoutines.cpp#L1107), the `x` and `y` have a size of `0`.
This makes the [next line](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Helmholtz/VLERoutines.cpp#L1108) to crash as it wants to [put a value](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Helmholtz/VLERoutines.cpp#L1093) in that vector and `K`is of size 3 (as I have a 3 component mixture in my test).

**Comment by JonWel** [comment]
In the cubic [set_mole_fractions](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Cubics/CubicBackend.h#L46), isn't it required to also resize the mole fraction vector for the `SatL` and `SatV`:
```C++
    if (this->SatL) this->SatL->resize(N);
    if (this->SatV) this->SatV->resize(N);
```
Like in the [HelmholtzEOSMixtureBackend](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Helmholtz/HelmholtzEOSMixtureBackend.cpp#L117)?

**Comment by JonWel** [comment]
Tranforming the cubic [set_mole_fractions](https://github.com/CoolProp/CoolProp/blob/master/src/Backends/Cubics/CubicBackend.h#L46) from
```C++
    void set_mole_fractions(const std::vector<CoolPropDbl> &mole_fractions){
        resize(mole_fractions.size());
        this->mole_fractions = mole_fractions; 
        this->mole_fractions_double = std::vector<double>(mole_fractions.begin(), mole_fractions.end());
};
```
to:
```C++
    void set_mole_fractions(const std::vector<CoolPropDbl> &mole_fractions){
        std::size_t N = mole_fractions.size();
        resize(N);
        // Resize the vectors for the liquid and vapor,  but only if they are in use
        if (SatL) SatL->resize(N);
        if (SatV) SatV->resize(N);

        this->mole_fractions = mole_fractions; 
        this->mole_fractions_double = std::vector<double>(mole_fractions.begin(), mole_fractions.end());
    };
```
Solves the crash for me.

Should I modify it in the header or makes the bigger definition in the cpp file? (as it is in the HelmholtzEOSMixtureBackend)

The HelmholtzEOSMixtureBackend seems to have a `N` in that can be accessed here. In the Cubic header this variable was not present. Is this a correct idea to create a varriable, or is this value efficiently retrieved from `.size()` (thus better to use `mole_fractions.size()` 3 times)?

**Comment by JonWel** [comment]
In deed, only functions that are different from the HelmholtzEOSMixtureBackend should be added in the Cubic.
In this case, the `set_mole_fractions` from the `HelmholtzEOSMixtureBackend` could be used (however we need to initialise N somewhere in the cubic, `set_components` is not called in the cubic backend), and the one of the `AbstractCubicBackend` removed.

Do you suggest to use the for loop to apply `resize()` iteratively to all the linked states?
Maybe we can use it inside resize:
```C++
void HelmholtzEOSMixtureBackend::resize(std::size_t N)
{
    for (std::vector<shared_ptr<HelmholtzEOSMixtureBackend> >::iterator it = linked_states.begin(); it != linked_states.end(); ++it){
        this->mole_fractions.resize(N);
        this->K.resize(N);
        this->lnK.resize(N);
     }
}
```

**Comment by ibell** [comment]
The idea is that we have some states that are linked as pointers to the
main state, and sometimes, when you want to change something in the main
state, the same modification should be made in the linked states.  Examples
would be setting interaction parameters, mole fractions, etc.

I think it would be best to re-use the base class code as much as
possible.  Perhaps the best would be to initialize N in the initializer of
the AbstractCubic backend?

**Comment by JonWel** [comment]
So, the idea I currently have is:

- Removing the:
```C++
    if (this->SatL) this->SatL->resize(N);
    if (this->SatV) this->SatV->resize(N);
```
From the  `HelmholtzEOSMixtureBackend::set_mole_fractions`
- Change `HelmholtzEOSMixtureBackend::resize(std::size_t N)` to:
```C++
void HelmholtzEOSMixtureBackend::resize(std::size_t N)
{
    this->mole_fractions.resize(N);
    this->K.resize(N);
    this->lnK.resize(N);
    for (std::vector<shared_ptr<HelmholtzEOSMixtureBackend> >::iterator it = linked_states.begin(); it != linked_states.end(); ++it){
        it->get()->resize(N);
     }
}
```
- And set `N` at the beginning of the `PengRobinsonBackend` and `SRKBackend` constructors:
```C++
    N = fluid_identifiers.size();
```

The new resize still work with the HEOS backend, it loops two times, corresponding exactly to the two saturation states.

However, it does not loop at all for the cubic equations. I guess the `linked_states` are not yet working for the cubics.
I take a look.

**Comment by JonWel** [comment]
Yep:
```
linked_states.push_back(SatL);
```
Are missing in the cubic.

I think I have enough to prepare a commit now.

**Comment by JonWel** [comment]
It looks like `set_mole_fractions` is called a first time with everything going correctly, and a second time, with a `N = 0`.

Why is `set_mole_fractions` called more than once? (17 times on HEOS where N is always with the correct value, it stop after two in cubic because the second time `N` doesn't have the correct value).

**Comment by JonWel** [comment]
It now woks (The copy for the two saturation states do not copy the value for N.
So I also update N for the saturation state in the for loop.

I'm preparing a PR.

**Comment by JonWel** [comment]
Closed by 4d679c0
